### ISSUE TITLE
Problem with big messages parsing

### ISSUE THREAD

**Issue by patriotyk** [GENIVI/rvi_lib#31]
If we receive  message that is bigger, than allocated buffer size, rvi_lib don't receive next part of message in current  context, and don't concatenate it.  Instead, it tries to parse part of message which is ofcourse damaged json object, and as result we got json parsing error.

**Comment by tjamison** [comment]
PR #32 addresses this by using the OpenSSL BIO API to determine the pending message length. Please comment on whether this addresses your issue.

**Comment by cajun-rat** [comment]
I don't know OpenSSL or rvi_lib very well, so forgive me if I have got
this wrong.

My understanding of the bug is that it is caused by the process that
splits the stream of bytes from the SSL/TLS into JSON objects (which are
then passed to json_loads).

At the moment the code seems to assume that the result of BIO_read will
be a whole protocol frame (aka JSON object), but I can't see why that
will always be the case.  My understanding of OpenSSL is that it
provides an interface like TCP, where read() can return an arbitrary
number of bytes from the stream.

Does that match your understanding?

**Comment by patriotyk** [comment]
@cajun-rat  Yes You are right,  and there is nice explanation https://mta.openssl.org/pipermail/openssl-users/2015-December/002569.html. And at the end of answer there is in my opinion acceptable solution - 

> When you have a successful SSL_read, use SSL_pending to get the number of application-data bytes remaining. Allocate a buffer of fixed-small-buffer-size + value-from-SSL_pending. Copy in the small fixed buffer, then SSL_read into the tail of the allocated buffer.

**Comment by tjamison** [comment]
I see. After some experimenting with SSL_pending and SSL_read, I adopted a modified approach in PR #33. It's a preliminary approach (still has an arbitrary size cap of about 1.6Mb which I intend to fix later) but it tested properly with records in the 10s of Ks. (The underlying BIO is set to blocking with the SSL_MODE_AUTO_RETRY flag set, so per the [SSL docs](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_mode.html), I expect SSL_read to return only after a successful read of the maximum number of bytes available to read from the stream -- the behavior would be different with non-blocking i/o or with the SSL_MODE_ENABLE_PARTIAL_WRITE flag set.)

**Comment by cajun-rat** [comment]
What character is used in the rvi_lib json protocol to separate json frames?

**Comment by cajun-rat** [comment]
OK, it seems like the protocol requires a streaming JSON parser (i.e. the framer needs to parse the JSON and look for the closing '}')

> from Ulf Wiger to Everyone: The way the erlang-based JSON parser handles it is that it is continuation-based so it receives a packet and either returns a parsed structure + the remainder, or a continuation to be called when there is more data.

**Comment by tjamison** [comment]
Right. The current implementation relies on a couple of things:
- That the SSL_read behavior reads complete TLS frames
- That the sending node will send each RVI command (json object) in one (or more) TLS frames

... so that reading TLS frames one-by-one and parsing at the first opportunity will only result in one RVI command (json object) to be parsed

Note that the current fix relies on the assumptions that:
- The maximum read size SSL_read is the same on all platforms (semi-reasonable, 2^14 bytes is the maximum size permitted by the TLS specification and the OpenSSL library handles reconstructing TLS fragments, and the client is responsible for initiating a maximum fragment length negotiation under the TLS extensions defined in RFC6066)
- Only multi-part records will hit the maximum (true for the majority of cases, but obvious exceptions apply)

**Comment by tjamison** [comment]
@patriotyk @cajun-rat @codders:

PR #34 supports parsing large/multiple messages in a continuous style:
* If a message cannot be parsed into valid JSON, the message is stored in a buffer associated with that fd until the next read on the fd, in which case the messages are concatenated and another attempt to parse JSON is made. rviProcessInput will return RVI_ERR_JSON as an error value. When used with level-triggered polling (as in select()), if subsequent fragments of a large message are still pending, the fd will be available immediately for read.
* If a message can be parsed into valid JSON, the corresponding RVI command will be processed. If there was additional data read, only the additional portion will be stored in a buffer to be concatenated with future messages. Either way, rviProcessInput will return 0 for success.

**Comment by patriotyk** [comment]
@tjamison I think there is one problem, we don't know if json is corrupted or just truncated. So in case of corruption we will save this corrupted json in to memory, and continue reading and forever appending data to memory until consume all memory.

**Comment by uwiger** [comment]
@patriotyk, this shouldn't happen if the reentrant parser works properly. It must distinguish between 1) complete, valid JSON,  2) incomplete, but so far valid JSON, and 3) invalid data. In the case of 2) the parser saves its state and thus evaluates for each subsequent char, if it's still a potentially valid parse result.

Of course, the wait for additional data should not be indefinite.

**Comment by tjamison** [comment]
No, @patriotyk is right about this issue. Jansson doesn't provide an incremental parser as such (it's still an open feature request from 2010: https://github.com/akheron/jansson/issues/5).

A major issue here is that Jansson's error reporting doesn't distinguish between JSON strings that appear to be too short vs JSON strings that are complete but invalid. 

[jsmn](https://github.com/zserge/jsmn) provides error checking that distinguishes between these two conditions. Its overhead is minimal, but adopting it would mean two JSON library dependencies for rvi_lib: libJWT still requires Jansson.

Another option would be to require that the stream always begin on '{', and use a timeout or provide a method for the calling application to flush the buffer. (The buffer is already flushed when the client closes the connection.)

@patriotyk Is one of these solutions better than the other for SOTA purposes? Adding jsmn would handle more cases, but I'm aware of the need to keep dependencies down.

**Comment by patriotyk** [comment]
Hi @tjamison, I don't understand You about timeout, and flush method. 

But I think we just need to improve only two things in Your PR:
1) Limit message size(don't append buffer infinitely), and also we should be sure that the server will newer send messages bigger. So this size should be documented in the protocol reference.
2) Send  RVI_ERR_JSON only when the message size limit exceeded and we was not able parse message. Otherwise we should return new error state, something like RVI_ERR_PARTIAL_JSON

Or rewrite library to use jsmn, which will be more sophisticated solution.

**Comment by tjamison** [comment]
On investigation, jsmn is not a viable solution for the error reporting requested: the returned value does _not_ consistently distinguish between (1) complete, valid JSON; (2) incomplete, but so far valid, JSON; and (3) invalid JSON (https://github.com/zserge/jsmn/issues/105).

I'll implement the message size limit (and new error state) and discuss with @uwiger the client-server message size negotiation so that it's properly documented in the protocol reference. 

I'll also implement a method to ensure that the buffer starts on a JSON object boundary so that having some garbage characters at the start of the buffer (e.g., from the tail end of an over-large message) doesn't cause subsequent valid messages to be discarded.

**Comment by tjamison** [comment]
See https://github.com/GENIVI/rvi_lib/pull/36. Note that the client still checks the maximum message size (if defined) before reading incoming messages to ensure reasonable behavior if the server disregards the maximum message size negotiation. 

As the updated README indicates, if the server sends an over-large message, rviProcessInput will return RVI_ERR_JSON repeatedly until there is a valid RVI command or no more data remaining to be read.
